### PR TITLE
Add SRP and hash password functionality

### PR DESCRIPTION
**Pull request by mmso**
- [X] Add SRP
- [X] Add hash password

### CODE REVIEW

File: lib/message/utils.js [hunk 1]
```diff
@@ -132,18 +132,11 @@ export async function handleVerificationResult({ data, filename = 'msg.txt', sig
...
-        .verify(options)
-        .then((result) => handleVerificationResult(result, publicKeys, options.date))
-        .then(({ data, verified, signatures }) => ({ data, verified, signatures }))
-        .catch((err) => {
-            console.error(err);
-            return Promise.reject(err);
-        });
+export async function verifyMessage(options) {
+    const { publicKeys = [], date: optionDate } = options;
+    const date = typeof optionDate === 'undefined' ? serverTime() : optionDate;
```
Comment (dhoko) [inline lib/message/utils.js]: you can use a default value when you destructure the date ;)

File: lib/srp/srp.js [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+    if (subtracted.compare(ZERO_BN) < 0) {
+        subtracted = subtracted.add(modulus);
+    }
+
+    const exponent = scramblingParam
+        .multiply(hashedPassword)
+        .add(clientSecret)
+        .divide(modulus.subtract(ONE_BN)).remainder;
+    const sharedSession = modulus.power(subtracted, exponent);
+
```
Comment (dhoko) [inline lib/srp/srp.js]: What is this ?

File: lib/srp/utils.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import { getRandomValues } from '../crypto';
+import { verifyMessage } from '../message/utils';
+import { VERIFICATION_STATUS, SRP_MODULUS_KEY } from '../constants';
+
+const { NOT_SIGNED, SIGNED_AND_VALID } = VERIFICATION_STATUS;
+
+/**
+ * Get modulus keys.
+ * Can cache keys since the SRP_MODULUS_KEY is constant.
+ * @return {Promise}
```
Comment (dhoko) [inline lib/srp/utils.js]: Where is the cache ?

File: lib/srp/utils.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+/**
+ * Verify modulus from the API and get the value.
+ * @param {String} modulus - Armored modulus string
+ * @returns {Promise<Uint8Array>}
+ */
+export const verifyAndGetModulus = async (modulus) => {
+    const [publicKeys, modulusParsed] = await Promise.all([
+        getModulusKeys(), //
```
Comment (dhoko) [inline lib/srp/utils.js]: There is an empty comment ;)

File: package.json [hunk 5]
```diff
@@ -5,8 +5,8 @@
   "main": "lib/index.js",
   "module": "lib/pmcrypto.js",
   "scripts": {
-    "test": "ava -v",
-    "lint": "eslint $(find lib -type f -name '*.js')  --quiet",
+    "test": "mocha -r esm 'test/**/*.spec.js'",
+    "lint": "eslint $(find lib -type f -name '*.js') --quiet",
```
Comment (dhoko) [inline package.json]: You can add --cache for eslint ;) it would be faster

File: lib/passwords.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        // Remove bcrypt prefix and salt (first 29 characters)
+        return hash.slice(29);
+    }
+
+    // No salt, old-style
+    return password;
+};
+
+const hashPasswordVersion = {
+    4(password, salt, modulus) {
```
Comment (EpokK) [inline lib/passwords.js]: Can we give a **name for each key** instead of using a `Number`?

File: lib/srp/srp.js [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+/**
+ * Get the client secret. Loops until it finds a safe value.
+ * @param {Number} len
+ * @return {BigNumber}
+ */
+const getClientSecret = (len) => {
+    const comparator = BigNumber.fromNumber(len * 2);
+
+    while (true) {
```
Comment (EpokK) [inline lib/srp/srp.js]: This loop scare me.

**Pull request merged**